src/sage/matroids/circuits_matroid.pyx

```diff
@@ -0,0 +1,676 @@
+                raise ValueError("not an ordering of the groundset")
+
+        # compute broken circuits
+        cdef list BC = []
+        for C in self._C:
+            for e in ordering:
+                if e in C:
+                    BC.append(C - set([e]))
+                    break
+
+        cdef list F = []  # broken circuit complex facets
+        for B in self.bases():
+            flag = True
+            for bc in BC:
+                if bc <= B:
+                    flag = False
+                    break
+            if flag:
+                F.append(B)
+
+        from sage.topology.simplicial_complex import SimplicialComplex
+        return [frozenset(f) for f in SimplicialComplex(F).face_iterator()]
+
+    # properties
+
+    cpdef girth(self) noexcept:
+        r"""
+        Return the girth of the matroid.
+
+        The girth is the size of the smallest circuit. In case the matroid has
+        no circuits the girth is `\infty`.
+
+        EXAMPLES::
+
+            sage: matroids.Theta(10).girth()
+            3
+
+        REFERENCES:
+
+        [Oxl2011]_, p. 327.
+        """
+        return min(self._k_C, default=float('inf'))
+
+    cpdef is_paving(self) noexcept:
+        """
+        Return if ``self`` is paving.
+
+        A matroid is paving if each of its circuits has size `r` or `r+1`.
+
+        OUTPUT: boolean
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.catalog.Vamos())
+            sage: M.is_paving()
+            True
+        """
+        return self.girth() >= self.rank()
+
+    # verification
+
+    cpdef is_valid(self) noexcept:
+        r"""
+        Test if ``self`` obeys the matroid axioms.
+
+        For a matroid defined by its circuits, we check the circuit axioms.
+
+        OUTPUT: boolean
+
+        EXAMPLES::
+
+            sage: C = [[1, 2, 3], [3, 4, 5], [1, 2, 4, 5]]
+            sage: M = Matroid(circuits=C)
+            sage: M.is_valid()
+            True
+            sage: C = [[1,2], [1, 2, 3], [3, 4, 5], [1, 2, 4, 5]]
+            sage: M = Matroid(circuits=C)
+            sage: M.is_valid()
+            False
+            sage: C = [[3,6], [1, 2, 3], [3, 4, 5], [1, 2, 4, 5]]
+            sage: M = Matroid(circuits=C)
+            sage: M.is_valid()
+            False
+            sage: C = [[3,6], [1, 2, 3], [3, 4, 5], [1, 2, 6], [6, 4, 5], [1, 2, 4, 5]]
+            sage: M = Matroid(circuits=C)
+            sage: M.is_valid()
+            True
+            sage: C = [[], [1, 2, 3], [3, 4, 5], [1, 2, 4, 5]]
+            sage: M = Matroid(circuits=C)
+            sage: M.is_valid()
+            False
+            sage: C = [[1, 2, 3], [3, 4, 5]]
+            sage: M = Matroid(circuits=C)
+            sage: M.is_valid()
+            False
+        """
+        for i in self._k_C:
+            for j in self._k_C:
+                if i <= j:
```

Comment:
```suggestion
        sizes = sorted(self._k_C)
        for ii, i in enumerate(sizes):
            for jj,j in enumerate(sizes[ii+1:], start=ii+1):
```
Fewer checks and simplifies the logic IMO. Actually, you probably can check if the first size is zero and then skip it to further simplify the next lines. The `jj` will be used below.

Reply:
Seems more complicated to me. These checks are nothing compared to the algorithm's complexity.

Reply:
If you want to do something by triangularity, then I think you should do it as such. I agree that the checks are likely not large, there are a lot of them (approx $\binom{n}{2}$ here plus more later on). Although I find this code easier to read and understand with less indentation. (I also got the triangular order backwards; I updated my suggestion above to fix this.) If you really disagree, then please change this to
```python
if i > j:
    continue
```